text/0000-custom-model-classes.md

```diff
@@ -0,0 +1,225 @@
+}
+  
+interface RecordDataFor {
+  (identifier: RecordIdentifier): RecordDataWrapper
+}
+    
+class Store {
+  instantiateRecord(
+    identifier: RecordIdentifier, 
+    store: Store,
+    createRecordArgs: { [key: string]: any }, // args passed in to store.createRecord() and processed by recordData to be set on creation
+    recordDataFor: RecordDataFor, 
+    notificationManager: NotificationManager): unknown
+}
+```
+
+Instead of exposing the entire RecordData Class, we expose only the local facing methods and hide the server/adapter facing methods that the Model should not have access to, by giving a RecordData lookup method to `instantiateRecord`
+
+## Change Notification
+
+In order for the record class to learn about changes to the underlying data, it will be passed a `NotificationManager` as a constructor argument, which will allow it to
+subscribe and unsubscribe to notifications of underlying changes to the data. Once `RecordData` calls one of the notification methods, the notification manager will call 
+any registered callback for the given identifier, and pass in the type of the notification, allowing the record the opportunity to repull the data if needed. There are no guarantees around the timing of the notification callback being called after `RecordData` informs the store of changes. We expect that in a modern Ember app with tracked properties, this wouldn't be the common path for tracking changes. 
+
+```ts
+interface UnsubscribeToken {
+  unsubscribe(): void;
+}
+
+interface NotificationCallback {
+  (identifier: Identifier, notificationType: 'attributes' | 'relationships' | 'errors'  | 'meta'  | 'unload'): void;
+}
+    
+interface NotificationManager {
+  subscribe(identifier, NotificationCallback): UnsubscribeToken
+ }
+```
+
+## Exposing schema information
+
+We currently keep schema information on `DS.Model` class. In order to allow for custom Model implementations we need to allow lookup of schema info from the store. We already have specified a schema api that RecordData consumes: `attributesDefinitionFor` and `relationshipDefinitionFor`. We would define these methods publicly on the store for addons to use. **These methods are not ergonomic on purpose.** They match the current Record Data apis and are designed as a stepping stone on the way of having a better, user facing schema APIs. The schema info is currently primarily geared towards being used by the internal relationship handling layer, the serializer/adapter layers and the DebugAdapter. Schema methods support both static and dynamic schema computation. For static schemas, the method can always respond with a schema definition based on the type passed, and for dynamic changing schemas, it can look up the underlying data by the identifer which is passed in. We would also add a method called `isModelType`, which would return `true` if ED knew that a given string is a model type and `false` otherwise.
```

Comment:
> `isModelType`, which would return `true` if ED knew that a given string is a model type and `false` otherwise.

How would this be determined?  Is this intended to be a user extension point?

Reply:
Yup